src/util/small_map.h

```diff
@@ -0,0 +1,560 @@
+///
+/// In the following we give some data about the memory consumption of
+/// small_mapt compared to std::map, on a 64-bit Linux system with GNU STL. It
+/// was determined with valgrind/massif. We configured the maps as
+/// `small_mapt<uint64_t, uint32_t, Num=8>` and `std::map<uint8_t, uint64_t>`.
+/// Thus, indices to small_mapt can be in {0, ..., 7} for this configuration.
+///
+/// Below we give the memory consumption (in bytes) of small_mapt and std::map,
+/// both for when containing 8 elements, and when being empty. The value for
+/// "useful" indicates the number of bytes requested, the value for "total" also
+/// includes padding bytes.
+///
+/// small_mapt:
+/// - 8 elements:
+///   * useful:  80 B
+///   * total:  120 B
+/// - empty:
+///   * useful:  16 B
+///   * total:   32 B
+///
+/// std::map:
+/// - 8 elements:
+///   * useful: 432 B
+///   * total:  512 B
+/// - empty:
+///   * useful:  48 B
+///   * total:   64 B
+///
+/// \tparam T mapped type
+/// \tparam Ind unsigned integer type, used to map integer indices to internal
+///   indices that index into the memory block that stores the mapped values
+/// \tparam Num gives range of valid indices, i.e., the valid indices are {0,
+///   ..., Num-1}, must satisfy Num * num_bits(Num-1) + Num < sizeof(Ind) * 8,
+///   with num_bits(n) denoting the minimum number of bits required to represent
+///   integer n
+template <typename T, typename Ind = uint32_t, std::size_t Num = 8>
+class small_mapt
+{
+public:
+  static_assert(std::is_unsigned<Ind>::value, "");
+
+  typedef Ind index_fieldt;
+
+  friend void small_map_test();
+
+  small_mapt() : ind(0), p(nullptr)
+  {
+  }
+
+  // Memory
+
+  void copy(T *dest, const T *src, std::size_t n) const
+  {
+    for(std::size_t i = 0; i < n; i++)
+    {
+      new(dest + i) T(*(src + i));
+    }
+  }
+
+  T *allocate(std::size_t n) const
+  {
+    if(n == 0)
+      return nullptr;
+
+    T *mem = (T *)malloc(sizeof(T) * n);
+
+    if(!mem)
+      throw std::bad_alloc();
+
+    return mem;
+  }
+
+  T *allocate(T *ptr, std::size_t n) const
+  {
+    if(n == 0)
+      return nullptr;
+
+    T *mem = (T *)realloc(ptr, sizeof(T) * n);
+
+    if(!mem)
+      throw std::bad_alloc();
+
+    return mem;
+  }
+
+  small_mapt(const small_mapt &m) : ind(m.ind), p(m.p)
+  {
+    if(m.p == nullptr)
+    {
+      return;
+    }
+
+    std::size_t n = m.size();
+    INVARIANT(n > 0, "size is > 0 if data pointer is non-null");
+
+    p = allocate(n);
+
+    copy(p, m.p, n);
+  }
+
+  ~small_mapt()
+  {
+    if(p)
+    {
+      std::size_t n = size();
+
+      for(std::size_t i = 0; i < n; i++)
+      {
+        (p + i)->~T();
+      }
+
+      free(p);
+    }
+  }
+
+  // Derived config
+
+  static const std::size_t N_BITS = std::numeric_limits<index_fieldt>::digits;
+
+  static const std::size_t NUM = Num;
+
+  static_assert(NUM >= 2, "");
+
+// When we don't have constexpr
+#if !(defined(__GNUC__) || _MSC_VER >= 1900)
```

Comment:
I'd suggest propagating  the negation as above. And wouldn't it make sense to move the above "no-constexpr" template instantiations down here?

Reply:
Unfortunately it can't be moved down as then it would be contained within another template. And then we could only specialize the template if we specialize the outer template as well.

Reply:
Oh, I should have thought of this. Maybe add a comment that this is used within the class? Reading the code top-to-bottom this piece of code comes a bit as a surprise.